File: student_management_system/student_class.py
import hashlib
import bisect


class Student:
    all_students = []

    def __init__(self, first_name, last_name, age, gender, school, cca, marks, phone_number):
        self.name = f"{first_name} {last_name}"
        self.age = age
        self.gender = gender
        self.school = school
        self.cca = cca
        self.marks = marks
        self.overall_marks = self.calculate_overall_marks()
        self.phone_number = phone_number
        Student.all_students.append(self)

    def print_details(self):
        print(f"Student ID: {self.generate_student_ID()}")
        print(f"Name: {self.name}")
        print(f"Age: {self.age}")
        print(f"Gender: {self.gender}")
        print(f"School: {self.school}")
        print(f"Co-Curricular Activities: {self.cca}")
        print(f"Marks: {self.marks_str()}")
        print(
            f"Performance (Best to Worst): {self.display_subject_descending()}")
        print(f"Overall Marks: {self.overall_marks}")
        print(f"Overall Grade: {self.calculate_overall_grade()}")
        print(f"Ranking: {self.calculate_ranking()}")
        print(f"Email: {self.email()}")
        print(f"Phone Number: {self.phone_number}")
        print("\n-------------------------------\n")

    def update_student_details(self):
        self.overall_marks = self.calculate_overall_marks()
        self.grade = self.calculate_overall_grade()
        self.ranking = self.calculate_ranking()

    def generate_student_ID(self):
        string = (self.name + self.school).lower()
        student_id = hashlib.sha256(string.encode()).hexdigest()[:8]
        return student_id

    def email(self):
        return f"{self.name.lower().replace(' ', '_')}@gmail.com"

    def marks_str(self):
        return ', '.join([f"{subject}: {mark}" for subject, mark in self.marks.items()])

    def display_subject_descending(self):
        sorted_subjects = sorted(
            self.marks.keys(), key=lambda x: self.marks[x], reverse=True)
        return ', '.join(sorted_subjects)

    def calculate_overall_marks(self):
        return round(sum(self.marks.values()) / len(self.marks))

    def calculate_overall_grade(self):
        grade_boundaries = [0, 40, 45, 50, 60, 70]
        grade_mapping = "FEDCBA"
        grade_index = bisect.bisect_right(grade_boundaries, self.overall_marks)
        return grade_mapping[grade_index - 1]

    def calculate_ranking(self):
        sorted_students = sorted(
            Student.all_students, key=lambda x: x.overall_marks, reverse=True)
        return sorted_students.index(self) + 1

File: Javascript/student_management_system/studentClass.js
import { createHash } from "crypto";

class Student {
    static allStudents = [];

    constructor(
        firstName,
        lastName,
        age,
        gender,
        school,
        cca,
        marks,
        phoneNumber
    ) {
        this.firstName = firstName;
        this.lastName = lastName;
        this.name = `${firstName} ${lastName}`;
        this.age = age;
        this.gender = gender;
        this.school = school;
        this.cca = cca;
        this.marks = marks;
        this.overallMarks = this.calculateOverallMarks();
        this.phoneNumber = phoneNumber;
        Student.allStudents.push(this);
    }

    printDetails() {
        console.log(`Student ID: ${this.generateStudentID()}`);
        console.log(`Name: ${this.name}`);
        console.log(`Age: ${this.age}`);
        console.log(`Gender: ${this.gender}`);
        console.log(`School: ${this.school}`);
        console.log(`Co-Curricular Activities: ${this.cca}`);
        console.log(`Marks: ${this.marksStr()}`);
        console.log(
            `Performance (Best to Worst): ${this.displaySubjectDescending()}`
        );
        console.log(`Overall Marks: ${this.overallMarks}`);
        console.log(`Overall Grade: ${this.calculateOverallGrade()}`);
        console.log(`Ranking: ${this.calculateRanking()}`);
        console.log(`Email: ${this.email()}`);
        console.log(`Phone Number: ${this.phoneNumber}`);
        console.log("\n-------------------------------\n");
    }

    updateStudentDetails() {
        this.overallMarks = this.calculateOverallMarks();
        this.grade = this.calculateOverallGrade();
        this.ranking = this.calculateRanking();
    }

    generateStudentID() {
        const string = (this.name + this.school).toLowerCase();
        return createHash("sha256").update(string).digest("hex").slice(0, 8);
    }

    email() {
        return `${this.name.toLowerCase().replace(" ", "_")}@gmail.com`;
    }

    marksStr() {
        return Object.entries(this.marks)
            .map(([subject, mark]) => `${subject}: ${mark}`)
            .join(", ");
    }

    displaySubjectDescending() {
        return Object.keys(this.marks)
            .sort((a, b) => this.marks[b] - this.marks[a])
            .join(", ");
    }

    calculateOverallMarks() {
        const totalMarks = Object.values(this.marks).reduce((a, b) => a + b, 0);
        return Math.round(totalMarks / Object.keys(this.marks).length);
    }

    calculateOverallGrade() {
        const gradeBoundaries = [0, 40, 45, 50, 60, 70];
        const gradeMapping = "FEDCBA";
        const gradeIndex = gradeBoundaries.findIndex(
            (boundary) => this.overallMarks < boundary
        );
        return gradeMapping[
            gradeIndex === -1 ? gradeMapping.length - 1 : gradeIndex - 1
        ];
    }

    calculateRanking() {
        const sortedStudents = Student.allStudents.sort(
            (a, b) => b.overallMarks - a.overallMarks
        );
        return sortedStudents.indexOf(this) + 1;
    }

    toDict() {
        return {
            name: this.name,
            age: this.age,
            gender: this.gender,
            school: this.school,
            cca: this.cca,
            marks: this.marks,
            phoneNumber: this.phoneNumber,
        };
    }

    static fromDict(data) {
        const [firstName, lastName] = data.name.split(" ");
        return new Student(
            firstName,
            lastName,
            data.age,
            data.gender,
            data.school,
            data.cca,
            data.marks,
            data.phoneNumber
        );
    }
}

export default Student;
